apsuite/commisslib/measure_tbt_data.py

```diff
@@ -0,0 +1,209 @@
+
+
+    def _get_tune_guess(self, matrix):
+        """."""
+        matrix_dft, tune = self.calc_spectrum(matrix, fs=1, axis=0)
+        tunes = tune[:, None] * _np.ones(matrix.shape[-1])[None, :]
+        peak_idcs = _np.abs(matrix_dft).argmax(axis=0)
+        tune_peaks = [tunes[idc, col] for col, idc in enumerate(peak_idcs)]
+        return _np.mean(tune_peaks)  # tune guess
+
+    def _get_amplitudes_and_phases_guess(self, matrix, tune):
+        """Calculate initial guesses for harmonic TbT model as
+           Eqs. (5.2)-(5.4) of Ref. [2]"""
+        N = matrix.shape[0]
+        ilist = _np.arange(N)
+        cos = _np.cos(2 * _np.pi * tune * ilist)
+        sin = _np.sin(2 * _np.pi * tune * ilist)
+        C, S = _np.dot(cos, matrix), _np.dot(sin, matrix)
+        C *= 2/N
+        S *= 2/N
+        amplitudes = _np.sqrt(C**2 + S**2)
+        phases = _np.unwrap(_np.arctan2(C, S))
+        return amplitudes, phases
+
+    def harmonic_tbt_model(self, ilist, amplitude, tune, phase):
+        """Harmonic motion model for positions seen at a given BPM"""
+        return amplitude * _np.cos(2 * _np.pi * tune * ilist + phase)
+
+    def harmonic_tbt_model_vectorized(self, ilist, amplitude, tune, phase):
+        """Harmonic motion model for positions seen at a given BPM"""
+        return amplitude[None, :] * _np.cos(2 * _np.pi * tune * ilist[:, None] + phase[None, :])
+
+    def fit_harmonic_model(self, matrix, amp_guesses,
+                           tune_guess, phase_guesses):
+        """Fits harmonic TbT model to data.
+
+        Args:
+            matrix (N,M)-array: data matrix containing N turns and BPMs
+            amp_guesses M-array: amplitude guess for each BPM time-series
+            tune_guess (float): betatron tune guess
+            phase_guesses M-array: phase guess for each BPM
+
+        Returns:
+            params (3, M)-array: fitted ampliude, tune and phase for each BPM.
+                                each column corresponds to a BPM, each row
+                                corresponds to amplitudes, tune and phase
+        """
+        ilist = _np.arange(matrix.shape[0])
+        params = _np.zeros((3, matrix.shape[-1]))
+        for bpm_idx, bpm_data in enumerate(matrix.T):
+            p0=[amp_guesses[bpm_idx],tune_guess,phase_guesses[bpm_idx]]
+            popt, *_ = _curve_fit(f=self.harmonic_tbt_model,
+                                  xdata=ilist,ydata=bpm_data,
+                                  p0=p0)
+            params[:, bpm_idx] = popt
+        return params
+
+    def calculate_betafunc_and_action(self, amplitudes, nominal_beta):
+        """Calculates beta function and betatron action as in Eq. (9)
+            of Ref. [1]"""
+        action = _np.sum(amplitudes**4)
+        action /= _np.sum(amplitudes**2 * nominal_beta)
+        beta = amplitudes**2/action
+        return beta, action
+
+class TbTDataParams(_AcqBPMsSignalsParams):
+    """."""
+
+    def __init__(self, hkick=None, vkick=None):
```

Comment:
Allowing inputs is not common in our Params classes. Additionally, this is a potential bug. If the user don't set this optional argument, hkick and vkick will be None and then `prepare_pingers` will set None to the pingers strength, which I think will lead to an error. 

A possible option is that you can initialize `self._hkick = None` and at some point check it is still None, do nothing with the pingers.

Reply:
Thanks! Hadn't thought about this possibility. Removed these properties as inputs and added checkings to avoid the  None type to be set as the magnets strengths.